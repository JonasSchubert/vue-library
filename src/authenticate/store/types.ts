export const ActionTypes = {
  login: 'login',
  logout: 'logout',
  validateToken: 'validate-token'
};

export const GetterTypes = {
  error: 'error',
  isAdministrator: 'isAdministrator',
  isLoading: 'isLoading',
  isUser: 'isUser',
  loggedIn: 'loggedIn',
  token: 'token',
  userRoles: 'userRoles'
};

export const ModuleType = 'authenticate';

export const MutationTypes = {
  setError: 'setError',
  setIsLoading: 'setIsLoading',
  setToken: 'setToken',
  setUserRoles: 'setUserRoles'
};

export const RouteTypes = {
  authenticate: 'authenticate'
};
