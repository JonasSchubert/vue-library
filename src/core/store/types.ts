export const GetterTypes = {
  isAnythingLoading: 'isAnythingLoading'
};

export const MutationTypes = {
  decreaseRequests: 'decreaseRequests',
  increaseRequests: 'increaseRequests'
};
