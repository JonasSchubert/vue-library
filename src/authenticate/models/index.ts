export * from './authenticate-module-setup.d';
export * from './authenticate-state.d';
export * from './base-authenticate-state.d';
export * from './login-data.d';
export * from './login-response.d';
export * from './validate-token-data.d';
