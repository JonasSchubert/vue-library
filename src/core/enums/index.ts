export * from './edit-mode.enum';
