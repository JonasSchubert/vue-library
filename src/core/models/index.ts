export * from './base-state.d';
export * from './entity.d';
