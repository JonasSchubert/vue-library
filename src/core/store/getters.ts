import { GetterTree } from 'vuex';
import { BaseState } from '../models';
import { GetterTypes } from './types';

export const createGetters = <T extends BaseState>(): GetterTree<T, T> => ({
  [GetterTypes.isAnythingLoading]: (state: T): boolean => state.requests > 0
});
