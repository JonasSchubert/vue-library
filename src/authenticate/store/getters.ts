import { GetterTree } from 'vuex';
import { GetterTypes } from './types';
import { UserRole } from '../enums';
import { AuthenticateState } from '../models';

export const createGetters = <TRootState>(): GetterTree<AuthenticateState, TRootState> => ({
  [GetterTypes.error]: (state: AuthenticateState): Error | undefined => state.error,
  [GetterTypes.isAdministrator]: (state: AuthenticateState): boolean => state.userRoles.includes(UserRole.Administrator),
  [GetterTypes.isLoading]: (state: AuthenticateState): boolean => state.isLoading,
  [GetterTypes.isUser]: (state: AuthenticateState): boolean => state.userRoles.includes(UserRole.User),
  [GetterTypes.loggedIn]: (state: AuthenticateState): boolean => !!state.token,
  [GetterTypes.token]: (state: AuthenticateState): string | undefined => state.token,
  [GetterTypes.userRoles]: (state: AuthenticateState): UserRole[] => state.userRoles
});
