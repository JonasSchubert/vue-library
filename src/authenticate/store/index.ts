import { Module } from 'vuex';
import { AuthenticateModuleSetup, AuthenticateState } from '../models';
import { createActions } from './actions';
import { createGetters } from './getters';
import { createMutations } from './mutations';
import { createAuthenticateState } from './state';

export const createAuthenticateModule = <TRootState>({ cookieKeyAuthenticationToken, daysTilExpiredAuthenticationCookie }: AuthenticateModuleSetup): Module<AuthenticateState, TRootState> => ({
  namespaced: true,
  actions: createActions(),
  getters: createGetters(),
  mutations: createMutations({ cookieKeyAuthenticationToken, daysTilExpiredAuthenticationCookie }),
  state: createAuthenticateState()
});
