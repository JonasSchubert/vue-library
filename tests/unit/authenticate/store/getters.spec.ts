import { UserRole } from '@/authenticate/enums';
import { AuthenticateState } from '@/authenticate/models';
import { createGetters } from '@/authenticate/store/getters';
import { GetterTypes } from '@/authenticate/store/types';

describe('authenticate/getters', () => {
  let state: AuthenticateState;

  const getters = createGetters();

  beforeEach(() => {
    state = {
      error: {
        message: 'message',
        name: 'name'
      },
      isLoading: false,
      token: 'token',
      userRoles: [UserRole.Administrator]
    };
  });

  describe(`${GetterTypes.error}`, () => {
    test('should return error as expected', () => {
      // Arrange && Act
      const actual = getters[GetterTypes.error](state, getters, {} as AuthenticateState, {});

      // Assert
      expect(actual).toMatchSnapshot({
        message: 'message',
        name: 'name'
      });
    });
  });

  describe(`${GetterTypes.isAdministrator}`, () => {
    test('should return isAdministrator as expected', () => {
      // Arrange && Act
      const actual = getters[GetterTypes.isAdministrator](state, getters, {} as AuthenticateState, {});

      // Assert
      expect(actual).toBeTruthy();
    });
  });

  describe(`${GetterTypes.isLoading}`, () => {
    test('should return isLoading as expected', () => {
      // Arrange && Act
      const actual = getters[GetterTypes.isLoading](state, getters, {} as AuthenticateState, {});

      // Assert
      expect(actual).toBeFalsy();
    });
  });

  describe(`${GetterTypes.isUser}`, () => {
    test('should return isEditor as expected', () => {
      // Arrange && Act
      const actual = getters[GetterTypes.isUser](state, getters, {} as AuthenticateState, {});

      // Assert
      expect(actual).toBeFalsy();
    });
  });

  describe(`${GetterTypes.token}`, () => {
    test('should return token as expected', () => {
      // Arrange && Act
      const actual = getters[GetterTypes.token](state, getters, {} as AuthenticateState, {});

      // Assert
      expect(actual).toBe('token');
    });
  });

  describe(`${GetterTypes.userRoles}`, () => {
    test('should return userRoles as expected', () => {
      // Arrange && Act
      const actual = getters[GetterTypes.userRoles](state, getters, {} as AuthenticateState, {});

      // Assert
      expect(actual).toMatchSnapshot([UserRole.Administrator]);
    });
  });
});
