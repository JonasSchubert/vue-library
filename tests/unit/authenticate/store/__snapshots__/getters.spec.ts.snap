// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate/getters error should return error as expected 1`] = `
Object {
  "message": "message",
  "name": "name",
}
`;

exports[`authenticate/getters userRoles should return userRoles as expected 1`] = `
Object {
  "0": 2,
}
`;
